### PR TITLE
Add test for the buffered error response bodies fix

### LINKED ISSUE #23251: [BUG] Response stream getting disposed for HttpWebRequestTransport.cs
When you run .net461 the HttpWebRequestTransport will dispose of the response stream even when it is a memory stream.

A temporary fix for this was added in this PR, but we will need to identify why our current unit tests did not catch this and potentially share this logic between this class and Response.cs

https://github.com/Azure/azure-sdk-for-net/pull/23249

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/23251

### CODE REVIEW

File: sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs [hunk 1]
```diff
@@ -537,7 +537,7 @@ public PipelineResponse(string requestId, HttpResponseMessage responseMessage, S
             public override void Dispose()
             {
                 _responseMessage?.Dispose();
-                DisposeContentStreamIfNotBuffered();
+                DisposeStreamIfNotBuffered(_contentStream);
```
Comment (allenjzhang) [inline sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs:540]: Can you add `_contentStream = null;`

Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs:540]: That won't be correct. We need to leave the stream assigned.

Comment (allenjzhang) [reply sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs:540]: Once you have disposed the stream, shouldn't leave it as not null.

File: sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs [hunk 2]
```diff
@@ -238,8 +238,9 @@ async Task Connect()
             await Task.WhenAll(requests);
         }
 
-        [Test]
-        public async Task BufferedResponsesReadableAfterMessageDisposed()
+        [TestCase(200)]
+        [TestCase(404)]
```
Comment (christothes) [inline sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs:242]: what's significant about these two status codes?

Comment (pakrym) [reply sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs:242]: On .NET Framework the HttpWebRequest transport always pre-buffers the error responses. This causes our buffering policy to opt-out. There was a bug where we accidentally disposed the buffered body in this case that was fixed in https://github.com/Azure/azure-sdk-for-net/pull/23249.

This PR adds a test and cleans up the implementation a bit.

Comment (christothes) [reply sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs:242]: Makes sense - I was just curious why we chose to test these specific response codes and why any response code is relevant to the test.

**Review by christothes — APPROVED** [review]

**Pull request merged**